core/events/domain/event.py

```diff
@@ -358,36 +362,337 @@ def project(self, submission: Submission) -> Submission:
-        """Update metadata on a :class:`.Submission`."""
-        for key, value in self.metadata:
-            setattr(submission.metadata, key, value)
+        """Update the ACM classification on a :class:`.Submission`."""
+        submission.metadata.acm_class = self.acm_class
         return submission
 
+    def _valid_acm_class(self, submission: Submission) -> None:
+        """Check that the value is a valid ACM class."""
+        ptn = r"^[A-K]\.[0-9m](\.(\d{1,2}|m)(\.[a-o])?)?$"
+        for acm_class in self.acm_class.split(';'):
+            if not re.match(ptn, acm_class.strip()):
+                raise InvalidEvent(self, f"Not a valid ACM class: {acm_class}")
+
+    def _cleanup(self, value: str) -> str:
+        """Perform light cleanup."""
+        value = re.sub(r"\s+", " ", value).strip()
+        value = re.sub(r"\s*\.[\s.]*$", "", value)
+        value = re.sub(r"^ACM-class:\s+", "", value, flags=re.I)
+        value = value.replace(",", ";")
+        _value = []
+        for v in value.split(';'):
+            v = v.strip().upper().rstrip('.')
+            v = re.sub(r"^([A-K])(\d)", "\g<1>.\g<2>", v)
+            v = re.sub(r"M$", "m", v)
+            _value.append(v)
+        value = ";".join(_value)
+        return value
+
+
+@dataclass
+class SetJournalReference(Event):
+    """Update the journal reference of a submission."""
+
+    journal_ref: str = field(default='')
+
+    def __post_init__(self):
+        """Perform some light cleanup on the provided value."""
+        self.journal_ref = self._cleanup(self.journal_ref)
+
+    def validate(self, submission: Submission) -> None:
+        """Validate the journal reference value."""
+        submission_is_not_finalized(self, submission)
+        if not self.journal_ref:    # Blank values are OK.
+            return
+        self._no_disallowed_words(submission)
+        self._contains_valid_year(submission)
+
+    def project(self, submission: Submission) -> Submission:
+        """Update the journal reference on a :class:`.Submission`."""
+        submission.metadata.journal_ref = self.journal_ref
+        return submission
+
+    def _no_disallowed_words(self, submission: Submission) -> None:
+        """Certain words are not permitted."""
+        for word in ['submit', 'in press', 'appear', 'accept', 'to be publ']:
+            if word in self.journal_ref.lower():
+                raise InvalidEvent(self,
+                                   f"The word '{word}' should appear in the"
+                                   f" comments, not the Journal ref")
+
+    def _contains_valid_year(self, submission: Submission) -> None:
+        """Must contain a valid year."""
+        if not re.search(r"(\A|\D)(19|20)\d\d(\D|\Z)", self.journal_ref):
+            raise InvalidEvent(self, "Must include a valid year")
+
+    def _cleanup(self, value: str) -> str:
+        """Perform light cleanup."""
+        value = value.replace('PHYSICAL REVIEW LETTERS',
```

Comment:
Eek, didn't realize this was in the old code, but not surprised. Probably generalizable but seems OK to leave as-is.

Reply:
Yeah, there are lots of things like this. There are also a bunch of patterns hard-coded for admin checks (some of what's in `admin/`) -- when we get to that point, we should think about an interface where they can maintain their own library of patterns.